Provide an int64 version of Quantity that is much faster

smarterclayton opened the pull request:
Resolves #25019 by adding an int64 version of Quantity that is aware of overflow and can spill to the inf.Dec version.  Heavily optimized all the paths for allocation and efficiency.

```
BenchmarkQuantityString-8           10000000           175 ns/op          23 B/op          2 allocs/op
BenchmarkQuantityStringBinarySI-8   10000000           214 ns/op          32 B/op          2 allocs/op
BenchmarkQuantityMarshalJSON-8      10000000           156 ns/op          35 B/op          1 allocs/op
BenchmarkQuantityUnmarshalJSON-8    10000000           194 ns/op           2 B/op          1 allocs/op
BenchmarkParseQuantity-8            10000000           142 ns/op           0 B/op          0 allocs/op
BenchmarkCanonicalize-8             20000000            89.2 ns/op         3 B/op          0 allocs/op
BenchmarkQuantityCopy-8             20000000            55.0 ns/op        48 B/op          1 allocs/op
BenchmarkQuantityAdd-8              50000000            33.0 ns/op         0 B/op          0 allocs/op
BenchmarkQuantityRound-8            50000000            25.1 ns/op         0 B/op          0 allocs/op
BenchmarkScaledValueSmall-8         30000000            35.9 ns/op         0 B/op          0 allocs/op
BenchmarkScaledValueLarge-8          3000000           511 ns/op          48 B/op          1 allocs/op
```

Known issues
- [x] Does not respect the current 12E max value, since that can be represented as 12*10^18 without loss of precision
- [x] Shifting the scale down to merge the denominator with numerator in Parse (which avoids going to inf.Dec), i.e. "1.2" becomes "12e-1".  Questionable.
- [x] Implement protobuf as a string.

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/25243)

<!-- Reviewable:end -->

smarterclayton commented:
Impact to general conversion and encoding

```
$ benchcmp /tmp/old1 /tmp/new1
benchmark                                      old ns/op     new ns/op     delta
BenchmarkDecodeCodec-8                         74170         57099         -23.02%
BenchmarkDecodeCodecToInternalProtobuf-8       14317         10915         -23.76%
BenchmarkDecodeIntoExternalCodec-8             63496         51679         -18.61%
BenchmarkDecodeIntoInternalCodec-8             71734         57493         -19.85%
BenchmarkDecodeIntoJSON-8                      57149         51707         -9.52%
BenchmarkDecodeIntoJSONCodecGen-8              34977         28385         -18.85%
BenchmarkDecodeIntoProtobuf-8                  5680          4814          -15.25%
BenchmarkEncodeCodec-8                         21975         19642         -10.62%
BenchmarkEncodeCodecFromInternal-8             28122         23517         -16.38%
BenchmarkEncodeCodecFromInternalProtobuf-8     11683         10165         -12.99%
BenchmarkEncodeCodecProtobuf-8                 6601          5533          -16.18%
BenchmarkEncodeJSONMarshal-8                   23308         19384         -16.84%
BenchmarkEncodeProtobufGeneratedMarshal-8      5460          4606          -15.64%
BenchmarkNodeConversion-8                      9641          6050          -37.25%
BenchmarkNodeCopy-8                            4746          6109          +28.72%
BenchmarkPodConversion-8                       6875          5819          -15.36%
BenchmarkPodCopy-8                             4433          4627          +4.38%
BenchmarkReplicationControllerConversion-8     7181          7133          -0.67%
BenchmarkReplicationControllerCopy-8           4116          4194          +1.90%

benchmark                                      old allocs     new allocs     delta
BenchmarkDecodeCodec-8                         337            226            -32.94%
BenchmarkDecodeCodecToInternalProtobuf-8       136            82             -39.71%
BenchmarkDecodeIntoExternalCodec-8             276            204            -26.09%
BenchmarkDecodeIntoInternalCodec-8             338            227            -32.84%
BenchmarkDecodeIntoJSON-8                      168            96             -42.86%
BenchmarkDecodeIntoJSONCodecGen-8              257            185            -28.02%
BenchmarkDecodeIntoProtobuf-8                  68             53             -22.06%
BenchmarkEncodeCodec-8                         66             39             -40.91%
BenchmarkEncodeCodecFromInternal-8             97             61             -37.11%
BenchmarkEncodeCodecFromInternalProtobuf-8     79             46             -41.77%
BenchmarkEncodeCodecProtobuf-8                 47             20             -57.45%
BenchmarkEncodeJSONMarshal-8                   62             34             -45.16%
BenchmarkEncodeProtobufGeneratedMarshal-8      43             16             -62.79%
BenchmarkNodeConversion-8                      85             29             -65.88%
BenchmarkNodeCopy-8                            28             32             +14.29%
BenchmarkPodConversion-8                       54             36             -33.33%
BenchmarkPodCopy-8                             25             25             +0.00%
BenchmarkReplicationControllerConversion-8     38             38             +0.00%
BenchmarkReplicationControllerCopy-8           22             22             +0.00%
```

lavalamp commented:
I'd like to make a pass on this.

smarterclayton commented:
Note that I have no comments anywhere in the code (in case there was a desire to restructure), which I'll work on once someone takes a first pass and offers opinions.

wojtek-t commented:
OK - since @lavalamp wants to look into it anyway and I don't have strong opinions on it, I'm reassigning it to him for now.

lavalamp commented on pkg/api/helpers.go:
```diff
@@ -58,16 +58,10 @@ var Semantic = conversion.EqualitiesOrDie(
 		//
 		// Uninitialized quantities are equivalent to 0 quantities.
-		if a.Amount == nil && b.MilliValue() == 0 {
+		if a.Sign() == 0 && b.Sign() == 0 {
```
Why isn't the compare enough?

smarterclayton replied:
It is, except that I didn't know whether I wanted to treat nil
quantity as zero by default in compare.  Will change.

lavalamp commented:
Review status: 0 of 32 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/api/helpers.go, line 64 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHfyc4zigPIOg2OGdcL:-KHfyc4zigPIOg2OGdcM:1496331754) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/helpers.go#L64)):*
> ```Go
> 			return true
> 		}
> 		return a.Cmp(b) == 0
> ```

Still not sure why Cmp doesn't do the IsZero check for you?

---

*[pkg/api/resource/amount.go, line 97 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg-668VeERY7wboeVS:-KHg-669b0qQultOkFlo:1161416903) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/amount.go#L97)):*
> ```Go
> }
> 
> const mostNegative = -(mostPositive + 1)
> ```

Does this line do what you want it to? maybe just `~int64(0)`

---

*[pkg/api/resource/amount.go, line 126 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg1nZ0xXpD1kKp6v_N:-KHg1nZ0xXpD1kKp6v_O:-1761781385) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/amount.go#L126)):*
> ```Go
> }
> 
> func int64MultiplyScale(a int64, b int64) (int64, bool) {
> ```

Why is this different from the above?

---

*[pkg/api/resource/amount.go, line 159 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg1iLinYSYcV5mcFJr:-KHg1iLjvlZhmBb-6U3W:66501609) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/amount.go#L159)):*
> ```Go
> }
> 
> func int64MultiplyScale1000(a int64) (int64, bool) {
> ```

why do you need different methods for all of these?

---

*[pkg/api/resource/amount.go, line 318 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg3a1VOdGHX0UCMwFm:-KHg3a1WTrrWQir_nSiP:1317287140) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/amount.go#L318)):*
> ```Go
> }
> 
> func (a int64Amount) AsCanonicalBytes(out []byte) (result []byte, exponent int32) {
> ```

we really need to explicitly define this format if it's going to be our wire format.

---

*[pkg/api/resource/amount.go, line 344 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg3lRQ0zm_UfncorzX:-KHg3lRQ0zm_UfncorzY:-2013865738) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/amount.go#L344)):*
> ```Go
> }
> 
> func (a int64Amount) AsCanonicalBase1024Bytes(out []byte) (result []byte, exponent int32) {
> ```

Why would I call this? We shouldn't have two "canonical" byte formats.

---

*[pkg/api/resource/quantity.go, line 108 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg4Sf--ljVw6LEVDm_:-KHg4Sf--ljVw6LEVDma:12762177) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/quantity.go#L108)):*
> ```Go
> type CanonicalValue interface {
> 	AsCanonicalBytes(out []byte) ([]byte, int32)
> 	AsCanonicalBase1024Bytes(out []byte) ([]byte, int32)
> ```

maybe just pick one?

---

*[pkg/api/resource/quantity.go, line 150 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg6rfmP9nZ2uBQcluG:-KHg6rfmP9nZ2uBQcluH:564262970) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/quantity.go#L150)):*
> ```Go
> 	positive = true
> 	pos := 0
> 	end := len(str)
> ```

Cast to bytes for this function?

---

*[pkg/api/resource/quantity.go, line 577 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg60cW8Y2LTeI5Z0qf:-KHg60cW8Y2LTeI5Z0qg:-1739988925) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/quantity.go#L577)):*
> ```Go
> 		return out, nil
> 	}
> 	result := make([]byte, 18, 18)
> ```

how do you know 18 is a good number?

---

*[pkg/api/resource/quantity.go, line 579 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg5mK6bk-zl--oUtXv:-KHg5mK6bk-zl--oUtXw:-1155422261) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/quantity.go#L579)):*
> ```Go
> 	result := make([]byte, 18, 18)
> 	result[0] = '"'
> 	number, suffix := q.CanonicalizeBytes(result[1:1])
> ```

Can the Canonicalize function do an append? it's not obvious how this is working.

---

*[pkg/api/resource/quantity.go, line 640 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg5HLBfbH6eRlodGy9:-KHg5HLBfbH6eRlodGyA:711867058) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/quantity.go#L640)):*
> ```Go
> func NewScaledQuantity(value int64, scale Scale) *Quantity {
> 	return &Quantity{
> 		i:      int64Amount{value: value, scale: scale},
> ```

What if this is bigger than an int64?

---

*[pkg/api/resource/quantity.go, line 677 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg4mdTRRfyWib2PRgA:-KHg4mdTRRfyWib2PRgB:102603422) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/quantity.go#L677)):*
> ```Go
> // SetMilli sets q's value to be value * 1/1000.
> func (q *Quantity) SetMilli(value int64) {
> 	q.SetScaled(value, -3)
> ```

Keep named unit?

---

*[pkg/api/resource/quantity_proto.go, line 109 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg7PMDjpC7FF-M3U0P:-KHg7PMEHB-7dBm7CsJt:5173157) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/quantity_proto.go#L109)):*
> ```Go
> func (m *Quantity) Unmarshal(data []byte) error {
> 	l := len(data)
> 	iNdEx := 0
> ```

??

---

*[pkg/api/resource/quantity_test.go, line 31 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg7tv7mByHILoPGaiR:-KHg7tv8_tb_QE4jMhOi:551386628) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/api/resource/quantity_test.go#L31)):*
> ```Go
> var (
> 	testQuantityFlag = QuantityFlag("quantityFlag", "1M", "dummy flag for testing the quantity flag mechanism")
> 	useInfDec        = pflag.Bool("use-inf-dec", false, "If true, use inf.Dec for the tests")
> ```

would be great if the exact same tests could run against quantities constructed both ways, all the time.

---

*[pkg/kubectl/describe.go, line 1826 \[r1\]](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243#-KHg8QRAtJEz_IjGnRfc:-KHg8QRCP07KohG7d5pM:-694567373) ([raw file](https://github.com/kubernetes/kubernetes/blob/04a8515f5d4e6322dec300098c65b8eb8a3258f2/pkg/kubectl/describe.go#L1826)):*
> ```Go
> 			} else {
> 				value.Add(podReqValue)
> 				reqs[podReqName] = value
> ```

I'm troubled that this is necessary-- it means that quantity doesn't act like either a reference or a value. Should people be storing pointers now?

---


*Comments from [Reviewable](http://reviewable.kubernetes.io:80/reviews/kubernetes/kubernetes/25243)*
<!-- Sent from Reviewable.io -->

smarterclayton commented:
Is reviewable supposed to be working?  Dies and fails for me.

smarterclayton commented:
I'm waiting to see if reviewable starts working on monday - I think it's impressive that it's less useful now even than GitHub reviews.  Otherwise will respond point by point.

lavalamp commented:
> Is reviewable supposed to be working? Dies and fails for me.

I passed Tim on the way to go hide to do this code review and he turned it on just for me ;) (We're waiting on some red tape before turning it on for everyone.)

lavalamp commented:
It was kinda slow on my laptop so I'm not totally sold on it yet...

smarterclayton commented:
Since I can't see anything other than the left diff it makes it hard
to respond.  Will do a best guess.

lavalamp commented:
Ugh, that's annoying. Sorry, won't try mixed-method reviews again.

smarterclayton commented:
> Still not sure why Cmp doesn't do the IsZero check for you?

It does, will fix.

> Does this line do what you want it to? maybe just `~int64(0)`

That was cribbed from a comment from bradfitz on detecting overflow with multiplication in Go - using bit complement feels a bit ugly as well.

> func int64MultiplyScale(a int64, b int64) (int64, bool) {
>
> Why is this different from the above?

Scale doesn't have to check for negative b (added a comment and tests) which makes it faster in the hot loops.  Was ~10% savings in add and others.
 
> func int64MultiplyScale1000(a int64) (int64, bool) {
>
> why do you need different methods for all of these?

Commented and added tests, compiler is better at `* 1000` (significantly so).

> we really need to explicitly define this format if it's going to be our wire format.

Will add to the quantity comment.

> 
> func (a int64Amount) AsCanonicalBase1024Bytes(out []byte) (result []byte, exponent int32) {
> ```

This is called when you call `Canonicalize` on a `BinarySI`.  Extracted directly from previous impl.

> type CanonicalValue interface {
> 	AsCanonicalBytes(out []byte) ([]byte, int32)
> 	AsCanonicalBase1024Bytes(out []byte) ([]byte, int32)
> Maybe just one of these

Need both unless we want to handle `Ki` differently.

> 	positive = true
> 	pos := 0
> 	end := len(str)
>
> Cast to bytes for this function?

I debated, but it would be a more invasive change.  The alternative was duplicating the function for bytes or string, but didn't seem worth it for the last allocation.

> 		return out, nil
> 	}
> 	result := make([]byte, 18, 18)
>
> how do you know 18 is a good number?

int64 is max 18 places, and the exponent is usually 1 or 2.  Extracted to a constant with comments.

> 	result := make([]byte, 18, 18)
> 	result[0] = '"'
> 	number, suffix := q.CanonicalizeBytes(result[1:1])
>
> Can the Canonicalize function do an append? it's not obvious how this is working.

Yeah, that's why the check for `&number[0] == &result[1]` is there.  Added a comment.

> func NewScaledQuantity(value int64, scale Scale) *Quantity {
> 	return &Quantity{
> 		i:      int64Amount{value: value, scale: scale},
> What if this is bigger than an int64?

int64Amount can handle anything where the numerator has less precision than int64, and can be unconverted to dec.

> // SetMilli sets q's value to be value * 1/1000.
> func (q *Quantity) SetMilli(value int64) {
> 	q.SetScaled(value, -3)
> ```
> Keep named unit?

Yeah, will change.

> func (m *Quantity) Unmarshal(data []byte) error {
> 	l := len(data)
> 	iNdEx := 0
>
> ??

This was copied from generated code.  Will add comments to that effect (note // BEGIN ... sections inline).

> would be great if the exact same tests could run against quantities constructed both ways, all the time.

Did this in more places.  Could not figure out how to define dynamic functions effectively.

> 			} else {
> 				value.Add(podReqValue)
> 				reqs[podReqName] = value
> I'm troubled that this is necessary-- it means that quantity doesn't act like either a reference or a value. Should people be storing pointers now?

Fun fact, this code was broken before if you were adding different formats (because the format wouldn't be copied according to the behavior).  I think we should be treating these like quasi values for now, but in 1.4 or 1.5 make a decision on whether we want to do int128

smarterclayton commented:
> we really need to explicitly define this format if it's going to be our wire format.

Actually, we already do.

lavalamp commented:
> That was cribbed from a comment from bradfitz on detecting overflow with multiplication in Go - using bit complement feels a bit ugly as well.

hm, I copy-pasted it into the golang playground and it actually gave a compile error :/ maybe you just can't directly send it through fmt.Printf.

smarterclayton commented:
@k8s-bot test this issue #IGNORE (bot lost track)

lavalamp commented on pkg/api/resource/quantity_proto.go:
```diff
@@ -17,62 +17,287 @@ limitations under the License.
-func (q *Quantity) QuantityProto() *QuantityProto {
-	if q == nil {
-		return &QuantityProto{}
+func (m *Quantity) Marshal() (data []byte, err error) {
```
Document the format of the returned data-- I take it it's just a string in the same format as the json output?

smarterclayton replied:
I believe the godoc in the IDL covers it, but if not I will add.

On May 16, 2016, at 7:35 PM, Daniel Smith <notifications@github.com> wrote:

In pkg/api/resource/quantity_proto.go
<https://github.com/kubernetes/kubernetes/pull/25243#discussion_r63444003>:

>
> -// ProtoTime returns the Time as a new ProtoTime value.
> -func (q *Quantity) QuantityProto() *QuantityProto {
> -	if q == nil {
> -		return &QuantityProto{}
> +func (m *Quantity) Marshal() (data []byte, err error) {

Document the format of the returned data-- I take it it's just a string in
the same format as the json output?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/25243/files/2a61025932c283f19dcc4e003cc623c57cea71e0#r63444003>

lavalamp commented on pkg/api/resource/quantity_proto.go:
```diff
@@ -17,62 +17,287 @@ limitations under the License.
-	return p
+	return n
 }
+func sozGenerated(x uint64) (n int) {
```
unreferenced?

smarterclayton replied:
Probably.

lavalamp commented on pkg/api/resource/quantity_proto.go:
```diff
@@ -17,62 +17,287 @@ limitations under the License.
+	return n
+}
+
+func sovGenerated(x uint64) (n int) {
```
maybe make this name reasonable, since it's unlikely we'll be rerunning the code generator?

smarterclayton replied:
I'm concerned about changing anything, in that it's exactly like the
rest of our proto.  But I can.

lavalamp replied:
up to you.

lavalamp commented on pkg/api/resource/quantity.go:
```diff
@@ -287,7 +294,7 @@ func ParseQuantity(str string) (Quantity, error) {
 	case DecimalExponent, DecimalSI:
 		scale = exponent
-		precision = 15 - int32(len(num)+len(denom))
+		precision = 18 - int32(len(num)+len(denom))
```
should this 18 be the constant you made?

smarterclayton replied:
Probably a different one - "maxDecimalPlacesInBase10Int64".

On Mon, May 16, 2016 at 7:57 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/api/resource/quantity.go
> <https://github.com/kubernetes/kubernetes/pull/25243#discussion_r63446356>
> :
>
> > @@ -287,7 +294,7 @@ func ParseQuantity(str string) (Quantity, error) {
> >  	switch format {
> >  	case DecimalExponent, DecimalSI:
> >  		scale = exponent
> > -		precision = 15 - int32(len(num)+len(denom))
> > +		precision = 18 - int32(len(num)+len(denom))
>
> should this 18 be the constant you made?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25243/files/f8f37c149266b823a4a54d1b0efc21bf8376080b..2a61025932c283f19dcc4e003cc623c57cea71e0#r63446356>
>

smarterclayton replied:
Will iterate tonight, squash, and reorganize the commits.  I think I have
tests for all the code paths covered (specifically the Add / Sub stress
test), Cmp, and the existing String tests.

On Mon, May 16, 2016 at 8:17 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Probably a different one - "maxDecimalPlacesInBase10Int64".
>
> On Mon, May 16, 2016 at 7:57 PM, Daniel Smith <notifications@github.com>
> wrote:
>
>> In pkg/api/resource/quantity.go
>> <https://github.com/kubernetes/kubernetes/pull/25243#discussion_r63446356>
>> :
>>
>> > @@ -287,7 +294,7 @@ func ParseQuantity(str string) (Quantity, error) {
>> >  	switch format {
>> >  	case DecimalExponent, DecimalSI:
>> >  		scale = exponent
>> > -		precision = 15 - int32(len(num)+len(denom))
>> > +		precision = 18 - int32(len(num)+len(denom))
>>
>> should this 18 be the constant you made?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/25243/files/f8f37c149266b823a4a54d1b0efc21bf8376080b..2a61025932c283f19dcc4e003cc623c57cea71e0#r63446356>
>>
>
>

lavalamp commented:
ok, I think no more comments. LGTM

smarterclayton committed:
Add an int64 quantity implementation

Provides fast transformations of int64 scaled values with overflow to
inf.Dec as needed. Attempts to reduce allocations and expensive
operations in the critical path.

Alters the public signature of Quantity to encourage more efficient use

smarterclayton committed:
Update use of Quantity in other classes

smarterclayton commented:
Squashed down to two.

smarterclayton committed:
Handle resource.Quantity changes in eviction thresholds

smarterclayton commented:
@derekwaynecarr please review last commit, it updates threshold

derekwaynecarr commented on pkg/kubelet/eviction/types.go:
```diff
@@ -88,7 +88,7 @@ type statsFunc func(pod *api.Pod) (statsapi.PodStats, bool)
 
 // signalObservations maps a signal to an observed quantity
-type signalObservations map[Signal]resource.Quantity
+type signalObservations map[Signal]*resource.Quantity
```
this changed for convenience?  i assume there was no issue using quantity in a map value.

smarterclayton replied:
Convenience, avoided 20 more refactors elsewhere in the code and tests.

derekwaynecarr commented:
@smarterclayton - the last commit looks good, I will take a pass this afternoon to ensure that map that keyed on Threshold can key on a better key.

smarterclayton commented:
@k8s-bot test this issue #IGNORE

wojtek-t commented:
@smarterclayton - it failed with some strange error, can you please take a look?

smarterclayton commented:
Going to open a flake when I get to the computer - it looked like ppc64 had
a bad library dependency.

On May 19, 2016, at 8:00 AM, Wojciech Tyczynski <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton> - it failed with some
strange error, can you please take a look?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-220303717>

smarterclayton commented:
```
+++ [0519 04:51:15] linux/ppc64le: go build started
# k8s.io/kubernetes/cmd/hyperkube
/usr/local/go/pkg/tool/linux_amd64/link: running powerpc64le-linux-gnu-gcc failed: exit status 1
/tmp.k8s/go-link-284905442/go.o: In function `type..hash.[13]k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd.subsystem':
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go:1:(.text+0x2005770): relocation truncated to fit: R_PPC64_REL24 (stub) against `runtime.interhash'
/tmp.k8s/go-link-284905442/go.o: In function `type..eq.[13]k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd.subsystem':
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go:1:(.text+0x2005858): relocation truncated to fit: R_PPC64_REL24 (stub) against `runtime.ifaceeq'
/tmp.k8s/go-link-284905442/go.o: In function `type..hash.struct { a string; b k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/configs.Operator }':
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/seccomp/config.go:1:(.text+0x2005f90): relocation truncated to fit: R_PPC64_REL24 (stub) against `runtime.strhash'
/tmp.k8s/go-link-284905442/go.o: In function `type..hash.struct { a string; b k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/configs.Action }':
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/opencontainers/runc/libcontainer/seccomp/config.go:1:(.text+0x2006330): relocation truncated to fit: R_PPC64_REL24 (stub) against `runtime.strhash'
collect2: error: ld returned 1 exit status
```

Nothing in this PR should have touched opencontainer as a dependency.

smarterclayton commented:
https://www.technovelty.org/c/relocation-truncated-to-fit-wtf.html

Implication is suddenly we are too large than what our library expected?  Which is possible in hyperkube, with more output.

smarterclayton commented:
@luxas you have the most recent name attached to PPC - are you familiar with the implications of this error?

smarterclayton committed:
Disable ppc64le for now

luxas commented:
Is it a flake?

luxas commented:
@k8s-bot test this please github issue: #IGNORE

luxas commented:
Chatted with @smarterclayton, and we have to disable `ppc64le` in this PR in order to get it merged, but as soon as it's working again, I'll enable it. This PR is a release blocker.

If IBM is serious with Kubernetes on PowerPC, they should focus on debugging this issue.
(After all, it was I that wrote the code that cross-compiles for ppc64le, when I heard there was interest in it. But I'm not going to _maintain it forever_ when I a) haven't _got a machine to test on_ and b) am not _paid to_ do it. **I will** however, create PRs towards making arm, arm64 and ppc64le a platform Kubernetes may run OOTB on, and cross-compile things (binaries, images, addons) so it'll work. _That's_ my hobby. I hope you all understand the difference between those two)

I've already spent ~3hrs trying to debug it, but no luck so far.
I tried:
 - To compile with `-mcmodel=medium` and `-mcmodel=large`
 - To update the `opencontainers/runc` to master

Here's the failure: https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-220310871

Cross-ref: #25730 #17981

/cc @bpradipt @seelam @pensu

smarterclayton commented:
Flake was #25844

wojtek-t commented:
OK - it seems we have tests passing now.

@smarterclayton - do you want to squash 3rd commit? Anything before re-applying lgtm label?

smarterclayton commented:
Naw, not going to squash.

smarterclayton commented:
Spawned https://github.com/kubernetes/kubernetes/issues/25886 for reenabling powerpc

smarterclayton commented:
Bumping since this has the powerpc disablement and blocks proto.

lavalamp commented:
Manual merge.

bpradipt commented:
@luxas, first of all let me take this opportunity to sincerely thank you for the work you did to add the PowerPC arch support.  Kubernetes on PowerPC is very important for IBM and we are ramping up the effort as well. Looking forward to your continued support and guidance.  We are looking into the linker issue and will keep the community posted.

wojtek-t commented:
Ref #8132

timstclair commented:
<vent> I just wanted to publicly acknowledge the horror that is circular vendor dependencies... I recognize there was no way to make this change without manually modifying the vendored appc package, but this now adds a bunch of etxra steps to updating Godeps. Can we move the resource package to its own repo? </vent>

smarterclayton commented:
We can.

On Fri, May 20, 2016 at 2:41 PM, Tim St. Clair <notifications@github.com>
wrote:

> I just wanted to publicly acknowledge the horror that is circular vendor
> dependencies... I recognize there was no way to make this change without
> manually modifying the vendored appc package, but this now adds a bunch of
> etxra steps to updating Godeps. Can we move the resource package to its own
> repo?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-220686553>
>

laboger commented:
I'm looking into the PPC64 linker error.  As mentioned above, the problem occurs when building the hyperkube binary.  The intermediate go.o file that is generated contains a single text segment that is too large (>2**24 bytes). 

An interesting question was brought up above that relates to this problem.  This file seems to be pulling in many, many packages.  Has anyone investigated whether or not they are all really needed?  Unfortunately you are hitting this error on ppc64le but it indicates that even when the programs are successfully built maybe they are larger than they need to be.  Also, as new code gets developed the programs will continue to grow in size and you will probably at some point hit linker overflow errors like this on other platforms.

smarterclayton commented:
Kube is probably one of the largest Go ecosystem projects (juju is larger,
and openshift because it depends on Kube).

Honestly, I expect us to grow to at least 2x our current package set size
over the next year - there will certainly be some time when we try to
reduce dependencies, but I don't see that happening soon and we'd prefer to
just fix the problems in the compile chains.

On Wed, May 25, 2016 at 10:23 AM, Lynn Boger <notifications@github.com>
wrote:

> I'm looking into the PPC64 linker error. As mentioned above, the problem
> occurs when building the hyperkube binary. The intermediate go.o file that
> is generated contains a single text segment that is too large (>2**24
> bytes).
>
> An interesting question was brought up above that relates to this problem.
> This file seems to be pulling in many, many packages. Has anyone
> investigated whether or not they are all really needed? Unfortunately you
> are hitting this error on ppc64le but it indicates that even when the
> programs are successfully built maybe they are larger than they need to be.
> Also, as new code gets developed the programs will continue to grow in size
> and you will probably at some point hit linker overflow errors like this on
> other platforms.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-221591974>
>

laboger commented:
The plan is to fix the problem so that the GNU linker can handle the build of a golang program like this.  I was just wondering if the size of the program was a concern and if there were other ways to reduce the size.

smarterclayton commented:
It's not a concern per se - some folks have even made statements like
"we'll get to 1Gb binaries eventually" :).  I doubt we'll get that bad, but
200Mb Kube binaries seems possible.

On Wed, May 25, 2016 at 12:48 PM, Lynn Boger <notifications@github.com>
wrote:

> The plan is to fix the problem so that the GNU linker can handle the build
> of a golang program like this. I was just wondering if the size of the
> program was a concern and if there were other ways to reduce the size.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-221632938>
>

thockin commented:
Hey.  This is wrong.  This is totally and utterly wrong.  It makes our vendored code not reproducible.  I can no longer godep restore/save and get my original values back.  This has to roll back.

smarterclayton commented:
I believe upstream has fixed their circular dependency on us now, so we can
update and still compile.

On Fri, May 27, 2016 at 6:22 PM, Tim Hockin <notifications@github.com>
wrote:

> Hey. This is wrong. This is totally and utterly wrong. It makes our
> vendored code not reproducible. I can no longer godep restore/save and get
> my original values back. This has to roll back.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-222266077>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_pwatZAbUEIZbYCZXCpbfDZAsBOP_ks5qF26MgaJpZM4IYjvP>
> .
>

smarterclayton commented:
Correction - updated their circular dependency.

On Fri, May 27, 2016 at 6:24 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I believe upstream has fixed their circular dependency on us now, so we
> can update and still compile.
>
> On Fri, May 27, 2016 at 6:22 PM, Tim Hockin <notifications@github.com>
> wrote:
>
>> Hey. This is wrong. This is totally and utterly wrong. It makes our
>> vendored code not reproducible. I can no longer godep restore/save and get
>> my original values back. This has to roll back.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-222266077>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/ABG_pwatZAbUEIZbYCZXCpbfDZAsBOP_ks5qF26MgaJpZM4IYjvP>
>> .
>>
>
>

smarterclayton commented:
appc 0.8.4 has the change.

On Fri, May 27, 2016 at 6:25 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Correction - updated their circular dependency.
>
> On Fri, May 27, 2016 at 6:24 PM, Clayton Coleman <ccoleman@redhat.com>
> wrote:
>
>> I believe upstream has fixed their circular dependency on us now, so we
>> can update and still compile.
>>
>> On Fri, May 27, 2016 at 6:22 PM, Tim Hockin <notifications@github.com>
>> wrote:
>>
>>> Hey. This is wrong. This is totally and utterly wrong. It makes our
>>> vendored code not reproducible. I can no longer godep restore/save and get
>>> my original values back. This has to roll back.
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-222266077>,
>>> or mute the thread
>>> <https://github.com/notifications/unsubscribe/ABG_pwatZAbUEIZbYCZXCpbfDZAsBOP_ks5qF26MgaJpZM4IYjvP>
>>> .
>>>
>>
>>
>

thockin commented:
no, upstream for github.com/appc/spec/schema/types/isolator_resources.go is not fixed - ParseQuantity returns a value now, and they need a pointer

smarterclayton commented:
ab50d12e88f57788bf84b83fef2be236eb1fcc0b has it.  The latter reverted?

On Fri, May 27, 2016 at 6:29 PM, Tim Hockin <notifications@github.com>
wrote:

> no, upstream for github.com/appc/spec/schema/types/isolator_resources.go
> is not fixed - ParseQuantity returns a value now, and they need a pointer
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-222267081>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p_va1vdR3JDatNJBavrJJDNbOAtvks5qF3A8gaJpZM4IYjvP>
> .
>

thockin commented:
7da0fdd4dfce6fda81869ac6ecd6ef42247957ea undid it

smarterclayton commented:
I'm going to bump to https://github.com/kubernetes/kubernetes/pull/26459
ab50d12e88 until they pick up our next tag.

On Fri, May 27, 2016 at 6:42 PM, Tim Hockin <notifications@github.com>
wrote:

> 7da0fdd4dfce6fda81869ac6ecd6ef42247957ea undid it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25243#issuecomment-222268867>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p_LICO7GgOotyVAq9_a16MFylha6ks5qF3NmgaJpZM4IYjvP>
> .
>

smarterclayton commented:
Opened #26463 to verify that verify-godeps.sh is doing its job - pretty sure it was returning green on this PR the entire time, which is why we didn't fix this before it merged.
